docs(textual): custom `Msg` renderers (#12785)
cosmos/cosmos-sdk: docs/architecture/adr-050-sign-mode-textual.md

In particular, the newline `"\n"` (ASCII: 10) character is forbidden.

> ### 3. Strings MUST match the `^\*?(>* )?((?![\*>: ]).+: )?((?![\*>: ]).+)$` Regex

Author: @JimLarson Trying to word out the details we discussed about `*` `>` and ` `, I believe a Regex is way clearer.

Reviewer: What's the meaning of having multiple asterix (`*`) in the front? `\*`?

Author: You can't have multiple asterisks in the front with this Regex, see the examples of invalid strings.

Reviewer 2: Regexes can be problematic because different regex engines have different interpretations of the same regex expression. They are also not particularly human-friendly :) Normally things like this are specified in Backhaus-Naur Form (BNF), [here's a random example](https://datatracker.ietf.org/doc/html/rfc7233#section-2.1) from HTTP.

End of transaction messages
```

> ### 9. Custom `Msg`-Renderers

Author: I'll maybe add an example for custom `Msg`-renderers

Application developers may choose to not follow default renderer value output for their own `Msg`s. In this case, they can implement their own custom `Msg` renderer. This is similar to [EIP4430](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-4430.md), where the smart contract developer chooses the description string to be shown to the end user.

> This is done by setting the `cosmos.msg.v1.textual.custom_renderer` Protobuf option to `true`. This option CAN ONLY be set on a Protobuf message representing transaction message object (implementing `sdk.Msg` interface).

Reviewer 3: Is the intention to communicate this to clients with the proto option? Is so we should say this and also maybe have a descriptive proto field to explain the format

Author: Yes, it's mainly to ease communication between app devs and client devs. So we currently have these ideas:
1. Boolean: `custom_renderer = true`
2. An unique algorithm identifier: `custom_renderer = "agoric_msg_to_text_v1"` (proposed by @webmaster128)
3. Human readable format: `custom_renderer = "Username: <username_from_state>\nValue: <value>"` (if I understand your proposal @aaronc, this string is not meant to be parsed programatically, it's just for humans).

I went with bool initially, but slowly changing my mind towards some kind of human-readable string.

Reviewer 4: Given the concerns about the feature, consider requiring a string like "I know what I'm doing" or "Don't blame Cosmos-SDK". `:-)`

Author: I went with 2, but also proposing like Jim said a "I know what I'm doing" warning: https://github.com/cosmos/cosmos-sdk/pull/12785#discussion_r937985861

```proto
message MsgFooBar {
  option (cosmos.msg.v1.textual.custom_renderer) = true;

  // proto fields
}
```

> When this option is set to true on a `Msg`, the implementation CANNOT use the default value renderers to render the `Msg`. Instead, it MUST check that a custom `Msg`-renderer is implemented for `MsgFooBar` (e.g. registered in a registry), or else the implementation MUST throw an error or panic.

Reviewer 4: Let's not prevent the custom renderer from calling out to default renderers if it so chooses, e.g. there's no need to re-invent the wheel for dates. I'd say instead:

> When this option is set to true on a `Msg`, a registered function will transform the `Msg` into an array of one or more strings as described above (section number?), which MAY use the key/value format with the expert field prefix. These strings may be rendered from a `Msg` field using a default value renderer, or they may be generated from several fields using custom logic.

> When this option is set to true on a `Msg`, the implementation CANNOT use the default value renderers to render the `Msg`. Instead, it MUST check that a custom `Msg`-renderer is implemented for `MsgFooBar` (e.g. registered in a registry), or else the implementation MUST throw an error or panic.

Reviewer: ```suggestion
When this option is set to true on a `Msg`, the implementation CANNOT use the default value renderers to render the `Msg`. Instead, it MUST check that a custom `Msg` renderer is registered and implemented for `MsgFooBar`, otherwise we MUST return an error.
```

Reviewer: Let's avoid panics

> Moreover, the implementation must provide 2 functions: one for formatting from Protobuf to string, and one for parsing string to Protobuf. These 2 functions are provided by the application devloper. To satisfy point #1, these 2 functions MUST be bijective with each other as inverse.

Reviewer 4: I'd add that:

- The two data types being interconverted are a `Msg` using proto equality and an array of strings using Go's deep equality.
- Both functions accept the current context as an argument, allowing stored state to be read, e.g. account-specific settings which might affect the translations.
- Both functions SHOULD NOT modify state.
- Note that if state is modified in a way that affects the translation between the signature being created and the signature being verified, the transaction will be rejected.

Reviewer: Instead of:

> Both functions SHOULD NOT modify state.

I would say: renderer functions must be pure (must not have any side effect such us state modification or global parameter update).

Reviewer: Also we may be more clear that "for every `Msg m1, m2: m1.Equals(m2) <==> renderer.Format(m1) == renderer.Format(m2)`

Author: > I would say: renderer functions must be pure

Renderer functions should be able to read from state though, which make them unpure. To answer your question https://github.com/cosmos/cosmos-sdk/pull/12785#discussion_r936435378, it's why we pass the `ctx` too.

Reviewer: What's the use case for render function to read additional data from the state?

Author: For example a mapping in state from address to username/nickname. Then TEXTUAL would shown "mom" or "john" instead of "agoric1abcdef...."

Reviewer: Do we really want that? Wallet, to prepare that message will need to have an access to the latest state. Ideally, wallet should be able to create and sign transaction fully offline (eg by tracking locally the nonce and account number).

Author: Currently, even without custom renderers, wallets already need access to state for coin metadata. So I don't think it's a big addition here.

Reviewer: It's not needed to create and sign transaction. Only account nonce and number is needed - but they can track it by themselves and sync once for a while, or on request.

Reviewer 4: The custom renderer needs access to state at runTx time so it can generate the same petname substitutions.

Reviewer 2: > For example a mapping in state from address to username/nickname. Then TEXTUAL would shown "mom" or "john" instead of "agoric1abcdef...."

https://docs.microsoft.com/en-us/dotnet/standard/security/cryptographic-signatures

> Cryptographic digital signatures use public key algorithms to provide data integrity. When you sign data with a digital signature, someone else can verify the signature, and can prove that the data originated from you and was not altered after you signed it. 

If the signature of a message is based on anything other than the message itself, then the signature doesn't actually provide data integrity, and the signature isn't actually a signature.

> Renderer functions should be able to read from state though, which make them unpure.

Assuming renderer functions produce signable bytes, it is _essential_ that they are pure.

Reviewer 2: > For example a mapping in state from address to username/nickname. Then TEXTUAL would shown "mom" or "john" instead of "agoric1abcdef...."

This requires that username/nickname mappings are one-to-one with account addresses. Is that guaranteed by whatever system you'd be querying? If you could alias both `agoric1=mom` and `hacker1=mom` then transactions `from=me to=agoric1 amount=100` and `from=me to=hacker1 amount=100` would produce the same signable bytes. I hope the problems there are clear 😉

Reviewer 5: I also think offline signing is worth maintaining. The more state is required, the harder that gets. Denom metadata can easily be shipped to the user in a config as this hardly changes. A message renderer that accesses state was requested by a user but is not what I wish to build and something that should remain opt-in.

Regarding the bijectivity: I'm not convinced we use this property. The way we do it right now with Amino JSON is: create proto document, serialize document to proto bytes (A), generate Amino JSON from document (B). Now B is signed and A is submitted to chain. We don't need Amino JSON -> proto as part of the signature verification or signing (correct me if I'm wrong). Since A and B are different, it is perfectly possible to have a Amino JSON converter that changes meaning if implemented that way.

Reviewer 2: >  Since A and B are different, it is perfectly possible to have a Amino JSON converter that changes meaning if implemented that way.

Yes, and I hope it's clear why this is a problem.

Reviewer 5: What about this case: I sign `from=me to=agoric1 amount=100` assuming `agoric1=ABC`. But in the meantime the state changes to `agoric1=XYZ`. Now the chain made me sign something I did not want to sign.

Reviewer 5: > Yes, and I hope it's clear why this is a problem.

Assuming I have to trust the chain's code anyways, it is actully not that clear to me how it becomes a problem.

Reviewer: I was just going to write _There is one more problem with the state: it can change_. And before hitting send, I saw @webmaster128 comment.

There could be even more traps if we enable state access.

Reviewer 2: > Assuming I have to trust the chain's code anyways, it is actully not that clear to me how it becomes a problem.

...what?

You have to trust the chain's code to an extent. But message signing is one of many capabilities that decentralized/permissionless networks employ so that users _don't need_ to trust the chain, its implementation, or its participants.

The point of signing a message is to guarantee, in a cryptographic sense, the integrity of a specific sequence of bytes. That integrity is verifiable and trustable by anyone, no matter what intermediating users want to do. But only if you play by the rules! The rules say that, given a privkey/pubkey pair, signing and verifying look like this:

```
fn sign(msg, privkey) -> signature
fn verify(msg, signature, pubkey) -> bool
```

You sign a specific msg. A msg is a specific sequence of bytes. Not a domain object, not one of N possible renderings of a sequence of bytes, and so on. A signature for some msg M establishes the integrity of that msg M _only_.

If the Amino rendering of a msg transforms one of its fields in one way, and a textual rendering transforms that field (or a different field, or no fields) in a different way, then the renderings yield different byte sequences. So you could produce two different msgs that would have the same rendered form. If you swapped actual msg M1 out for a different msg M2 that rendered the same way, and then asked the signer to sign M2, you could then apply that signature to M1, and it would parse as valid. But the signer didn't sign M1, they signed M2. So you've lost integrity.

The signature of a msg/tx has to be on its canonical Merkelized bytes. You can't sign a rendering.

Author: > You have to sign the canonical, Merkelized bytes

The merklelized tx bytes include the tx signature, so by definition you can't sign the whole merklelized tx. What you sign contains the tx minus the signature.

> Signing a rendering doesn't provide integrity.

I argue that as long as `render_foo()` is 1/ part of consensus (so publicly and deterministically reproduceable by anyone who wants to verify sigs), and 2/ guarantees losslessness and non-malleability, then it provides integrity.

In all the SDK's sign modes, we sign a rendering of the tx (called `SignDoc*`).
- DIRECT: we sign the proto bytes of [this guy](https://github.com/cosmos/cosmos-sdk/blob/daca54d933d4a7c736324cc4add9f74d5394b648/proto/cosmos/tx/v1beta1/tx.proto#L48). Probably the closest to what's merklelized on chain, but still not quite that
- DIRECT_AUX: proto bytes of [this guy](https://github.com/cosmos/cosmos-sdk/blob/daca54d933d4a7c736324cc4add9f74d5394b648/proto/cosmos/tx/v1beta1/tx.proto#L71)
- AMINO_JSON: we sign the amino JSON representation of the tx

I'd go one step further than Jim's https://github.com/cosmos/cosmos-sdk/pull/12785#discussion_r943025204 to say that your concerns touch on existing sign modes too (especially the 2 latter). If in your eyes this is critical, SEV-0, red-alert, consider reporting this as a security issue, so that the relevant eyes takes a look at this quickly.

Reviewer 3: I think we should be signing the raw txbody and authinfo bytes in sign mode textual. They can be "advanced mode" strings, but if they're not there the tx is malleable

Author: After some more thinking, I came around on this, and I agree with @aaronc we should sign the hash of [body bytes](https://github.com/cosmos/cosmos-sdk/blob/daca54d933d4a7c736324cc4add9f74d5394b648/proto/cosmos/tx/v1beta1/tx.proto#L52) and [auth_info_bytes](https://github.com/cosmos/cosmos-sdk/blob/daca54d933d4a7c736324cc4add9f74d5394b648/proto/cosmos/tx/v1beta1/tx.proto#L56) as they are included on-chain. Or else we would have tx hash malleability (though the tx values are not malleable thanks to bijectivity).

I'll add a paragraph on this.

Reviewer 2: > I argue that as long as render_foo() is 1/ part of consensus (so publicly and deterministically reproduceable by anyone who wants to verify sigs), and 2/ guarantees losslessness and non-malleability, then it provides integrity.

Let's say we have a renderer that transforms addresses to nicknames, using on-chain "address books" which users can define for themselves, and which don't permit duplicate names. The renderer is part of consensus, the transformation is lossless and non-malleable. All points satisfied.

If user `agoric1` has addrbook `{agoric1=me agoric2=alice}` then the tx `{from=agoric1 to=agoric2 amount=100}` renders as `{from=me to=alice amount=100}` and produces some signature S1.

A malicious user `agoricX` can observe that tx and its rendering to create addrbook `{agoric1=me agoricX=alice}` and then tx `{from=agoric1 to=agoricX amount=100}` which would render as `{from=me to=alice amount=100}` and allow the malicious user to use the observed signature S1. I believe signature verification of this malicious tx would succeed?

This is just an example. Maybe other tx/msg properties would prevent this specific scenario. Maybe you could define "rules" for renderers like "don't transform sender addresses" or something. But I hope it's clear that the underlying issue remains. Signatures provide integrity only of the actual bytes signed, any transformation breaks the trust model.

Author: I added c3ac6edca6765e21f141d334bded76895f25dbfa, LMK (maybe in a new github thread) if that addresses your concerns.

Reviewer 4: [The example of the addressbook is back on-topic. :-)] The custom renderer must know to use the addressbook associated with the signer/sender for this transaction, so `agoricX` can construct whatever addressbook they'd like, but the rendering would use the addressbook of `agoric1`. It's not too much of a burden to expect the custom renderer to get this right - we already expect Msg implementors to implement `GetSigners()` correctly.

Reviewer 2: > The custom renderer must know to use the addressbook associated with the signer/sender for this transaction, so agoricX can construct whatever addressbook they'd like, but the rendering would use the addressbook of agoric1. It's not too much of a burden to expect the custom renderer to get this right - we already expect Msg implementors to implement GetSigners() correctly.

If msg implementors don't implement GetSigners correctly, I assume it will produce an error, or halt the chain, or in some way refuse to accept the bad data? If so, that's fine! But I've presented a categorically different situation. If msg implementors accidentally take the wrong addrbook from state, then it's possible for malicious users to fake signatures, and malicious data to enter the chain.

Reviewer 2: > The signature must be over data that includes the text presented to the user, otherwise malicious wallet software could send the wrong text to the Ledger and get the user to sign a completely different transaction.

There is an unavoidable trust assumption in any human-readable information derived from non-human-readable data. The text is always going to be model of the tx, not the tx itself. Cryptographic signatures operate on data, they aren't meant to be human-auditable, right? If this is something that the maintainers believe to be a requirement delivered by e.g. Ledger, then there must have been some amount of miscommunication or misunderstanding. It just don't make sense.

Reviewer 4: No - if, for instance, the bank "send" message messed up `GetSigners()` and specified that the recipient needed to sign rather than the sender, then the chain would happily let users drain each other's accounts. This method absolutely must be implemented correctly for the integrity of the chain. That's why I say that it's not a new burden to expect the renderer to select the right user's addressbook, and a screwup in doing so is not a new class of error.

Reviewer 4: The whole reason the Ledger has a screen is so that the user can understand what they are signing, and that only makes sense if the presented text cannot be manipulated without invalidating the signature.

Reviewer 2: > The whole reason the Ledger has a screen is so that the user can understand what they are signing, and that only makes sense if the presented text cannot be manipulated without invalidating the signature.

Cosmos transactions cannot be rendered in a way that's simultaneously human-readable, lossless, and non-malleable.

Reviewer 4: > Cosmos transactions cannot be rendered in a way that's simultaneously human-readable, lossless, and non-malleable.

Bummer. Someone better explain that to SIGN_MODE_LEGACY_AMINO_JSON.

Reviewer 2: JSON is arguably human-readable — it is for me! 😉 — but it is both malleable and lossy. At least, per the spec. if that sign mode uses a lossless and deterministic JSON encoder, then it may be OK!

But I did some research. SIGN_MODE_LEGACY_AMINO_JSON apparently (?) uses [tendermint/go-amino](https://github.com/tendermint/go-amino) to encode and decode, and I guess specifically the JSON methods in that package. I believe execution eventually gets to [Codec.encodeReflectJSON](https://github.com/tendermint/go-amino/blob/master/json-encode.go#L23), which forwards e.g. map encoding over to [encodeReflectJSONMap](https://github.com/tendermint/go-amino/blob/8e779b71f40d175cd1302d3cd41a75b005225a7a/json-encode.go#L343). That function iterates over the map using [reflect.Value.MapKeys](https://github.com/tendermint/go-amino/blob/8e779b71f40d175cd1302d3cd41a75b005225a7a/json-encode.go#L370), which [per documentation] (https://pkg.go.dev/reflect#Value.MapKeys)

> MapKeys returns a slice containing all the keys present in the map, **in unspecified order**. 

Can valid txs/msgs contain maps? I'm not sure, actually! But if they can, then not only is this sign mode lossy and malleable, it is also non-deterministic! :o

Author: Resolving this thread as off-topic. There's a security thread going on.

> Moreover, the implementation must provide 2 functions: one for formatting from Protobuf to string, and one for parsing string to Protobuf. These 2 functions are provided by the application devloper. To satisfy point #1, these 2 functions MUST be bijective with each other as inverse.

Reviewer: ```suggestion
Moreover, the renderer must provide 2 functions: one for formatting from Protobuf to string, and one for parsing string to Protobuf. These 2 functions are provided by the application devloper. To satisfy point #1, these 2 functions MUST be bijective with each other.
```

> Bijectivity of custom `Msg`-renderers SHOULD be checked with best effort by the SDK on node startup. To achieve this, we plan to use a similar technique to property testing, whereby a set of random inputs are generated, and passed into the relevant `Msg`. We then test that the composition of the two custom functions, formatting and parsing, result in the original Protobuf `Msg`:

Reviewer 4: I'd recommend that you either trust the custom rendering (and check bijectivity through provided tests) or don't (and check every time at runtime). I don't see the additional safety of checking at node startup time vs checking the software release - nothing should be different at node startup time.

Reviewer: Agree.

Author: The question here is: Should the Cosmos SDK check that the app dev's custom rendering is bijective?

> you either trust the custom rendering (and check bijectivity through provided tests)

How would the SDK check bijectivity through provided tests (which I suppose will live in `_test.go` files in the app dev's repo)?

Or the SDK checks nothing. But then this is really a footgun.

> or don't (and check every time at runtime)

The way I see it, the SDK will expose a `textual.RegisterCustomMsgRenderer()` to app chains. Probably somewhere in app.go, app devs will call it with their custom implementation. I was thinking to check bijectivity upon this function call, which happens at node startup.

We could also check bijectivity at runtime, i.e. when verifying a tx signature which uses TEXTUAL. But this will slow down performance, with no added benefit vs at node startup.

Reviewer 4: Yes, a comprehensive suite would be part of the app repo, and should be there regardless of where else checks might run.

The problem with a startup check is for applications like ours where the transformation depends on context, e.g. having a set of per-account petnames for addresses. It's easy to set up such context in a test and easy to check the real use case at runtime, but less straightforward and less representative of real use to check on startup. Do you make a real account on your mainnet just for testing? Now you've got to update this state if the tests change.

Because of these difficulties, I think a startup test would tend to be trivial, and a trivial test provides little protection compared to a comprehensive test suite, which would be inappropriate to run on startup.

Having the inverse mapping be part of the "Renderer" interface is a strong signal to developers for how this feature should be designed and tested. At some point you need to trust the developer, otherwise you won't have any powerful features.

Author: That's a good point. With state reads, the node startup check would indeed be trivial. I rephrased the ADR to say that the SDK won't check bijectivity, and it's the responsibility of the app developer to check it.

```

> parse(format(`MsgFooBar`)) == `MsgFooBar`

Reviewer 4: ```
// for renderer, msg, and ctx of the right type
keyvals, err := renderer.ToStrings(ctx, msg)
if err != nil {
    fail_check()
}
msg2, err := renderer.ToMsg(ctx, keyvals)
if err != nil {
    fail_check()
}
if !proto.Equal(msg, msg2) {
    fail_check()
}
pass_check()
```

Reviewer: Why do we need to pass `ctx`? I think it's not needed. Renderer functions must be pure.

Reviewer: Also, I wouldn't use `ToString` function. It implies that we want to format the renderer, rather than the argument. Instead, `Parse` and `Format` function names are widely used in Go APIs.

Reviewer 2: > Also, I wouldn't use `ToString` function. It implies that we want to format the renderer, rather than the argument. Instead, `Parse` and `Format` function names are widely used in Go APIs.

Format/Parse permit information loss. If that's not acceptable, you want Marshal/Unmarshal, or Encode/Decode.

End of transaction messages
```

### 9. Custom `Msg` Renderers

Application developers may choose to not follow default renderer value output for their own `Msg`s. In this case, they can implement their own custom `Msg` renderer. This is similar to [EIP4430](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-4430.md), where the smart contract developer chooses the description string to be shown to the end user.

This is done by setting the `cosmos.msg.v1.textual.custom_renderer` Protobuf option to a non-empty string. This option CAN ONLY be set on a Protobuf message representing transaction message object (implementing `sdk.Msg` interface).

```proto
message MsgFooBar {
  // Optional comments to describe in human-readable language the formatting
  // rules of the custom renderer.

>   option (cosmos.msg.v1.textual.custom_renderer) = "<unique algorithm identifier>";

Author: Given the discussions, I'm inclined to do:

```suggestion
  option (cosmos.msg.v1.textual.unsafe_custom_renderer) = "<unique algorithm identifier>";
```

ref: https://github.com/cosmos/cosmos-sdk/pull/12785#discussion_r937012747

WDYT?

Reviewer: TBH, I don't think `unsafe*` API should be exposed. It doesn't look good.

Reviewer 4: Agreed about "unsafe" - maybe 'expert" instead? Or just go with your original thought and make it a boolean. You'll want to register the renderer anyhow, so the algorithm identifier doesn't seem to be adding much.

Author: We agreed on our call to name this "expert"

  // proto fields
}
```

> When this option is set on a `Msg`, a registered function will transform the `Msg` into an array of one or more strings, which MAY use the key/value format (described in point #3) with the expert field prefix (described in point #5). These strings MAY be rendered from a `Msg` field using a default value renderer, or they may be generated from several fields using custom logic.

Author: ```suggestion
When this option is set on a `Msg`, a registered function will transform the `Msg` into an array of one or more strings, which MUST use the key/value format (described in point #3) and MAY use the expert field prefix (described in point #5). These strings MAY be rendered from a `Msg` field using a default value renderer, or they may be generated from several fields using custom logic.
```

any opinions on MAY v MUST here for the `key: value` format? If we do MUST, it'll probably make client UI lives easier.

Reviewer: +1 for MUST

Reviewer 4: Depends on what guarantees you want to make for the signing agent. I think the freedoms of a custom renderer should match the freedoms given the default renderer - the signing agent needs to handle either, and we can make things less complicated for it by not having a difference between the range of formats emitted by a custom vs default renderer.

Reviewer 4: After thinking about this more, the custom renderer must have the ability to generate multi-screen output, including aggregate start and end screens, so not everything will be key/value.

The `unique algorithm identifier` is a string convention chosen by the application developer and is used to identify the custom `Msg` renderer. For example, the SPEC documentation of this custom algorithm can reference this identifier. This identifier CAN have a versioned suffix (e.g. `_v1`) to adapt for future changes (which would be consensus-breaking). We also recommend adding Protobuf comments to describe in human language the custom logic used.

> Moreover, the renderer must provide 2 functions: one for formatting from Protobuf to string, and one for parsing string to Protobuf. These 2 functions are provided by the application developer. To satisfy point #1, these 2 functions MUST be bijective with each other. Bijectivity of these 2 functions will not be checked by the SDK at runtime. However, we strongly recommend the application developer to include a comprehensive suite in their app repo to test bijectivity, as to not introduce security bugs. A simple bijectivity test looks like:

Reviewer 2: A bijective rendering between two representations of a type is more accurately called an encoding. Do you actually intend that Cosmos should define, implementing, and maintain its own bespoke Protobuf encoding scheme?

Author: Yes. Hence the discussion [here](https://github.com/cosmos/cosmos-sdk/pull/12785#discussion_r939445681) about naming this custom encoding of Msgs "expert" or "unsafe".

Reviewer 2: Does that mean implementing this encoding for a given msg type is optional?

Author: Yes. By default, all Msgs are encoded with the default value renderers (see annex 1). If the app developer wishes to overwrite the default encoding, then it uses this custom Msg renderer.

Reviewer 2: Ah, okay! If a msg doesn't implement this textual/expert/unsafe renderer (encoder) will it still be sign-able on Ledger devices?

Reviewer 4: 1. Custom renderers are completely optional - Msgs with no custom renderer registered will use the default message rendering described in the annex.
2. I'd argue to include an option for runtime verification of bijectionality when custom extensions are used, but otherwise, there's a possibility that the signed text omits some important information. However, the risks of this must be weighed against the risks of illegibility from non-customized or non-textual presentations.
3. The Msg to text rendering for a given concrete Msg must be the same on the user's client software and on the validators, so the part that dispatches from the Msg, proto, or field value to the renderer must be deterministic.

Reviewer 2: > I'd argue to include an option for runtime verification of bijectionality when custom extensions are used, but otherwise, there's a possibility that the signed text omits some important information. However, the risks of this must be weighed against the risks of illegibility from non-customized or non-textual presentations.> The Msg to text rendering for a given concrete Msg must be the same on the user's client software and on the validators, so the part that dispatches from the Msg, proto, or field value to the renderer must be deterministic.

```
type MyMsg struct {
    Name string
    Desc string
}

func (m *MyMsg) RenderJSON() ([]byte, error) { return json.Marshal(m) }
func (m *MyMsg) RenderText() ([]byte, error) { return m.Name }
```

Reviewer 2: > The Msg to text rendering for a given concrete Msg must be the same on the user's client software and on the validators, so the part that dispatches from the Msg, proto, or field value to the renderer must be deterministic.

I'm not worried about the behavior of a given renderer, I'm worried about the notion that the signable bytes for a given tx/msg aren't well-defined but are instead produced by one of N arbitrary methods on the type. As an example, given

```
type MyMsg struct {
    Name string
    Desc string
}

func (m *MyMsg) RenderJSON() ([]byte, error) { return json.Marshal(m) }
func (m *MyMsg) RenderText() ([]byte, error) { return []byte(m.Name + "\n" + m.Desc) }
```

Is it the case that the signature for this msg could be based on either of these two methods depending on a parameter that's not part of the msg? If so, hopefully this example demonstrates why this design is unsound? 😬

Reviewer 4: There will need to be some method of registering a particular custom renderer and telling the dispatcher which message type it's supposed to handle. For instance, in Go you'd have something like the following in `app.go`:

```
textual := valuerenderer.NewTextual()
textual.RegisterCustom(NewMyMsgTextRenderer(), MyMsg{})
tx.EnableTextual(textual)
```
 
The exact details would be part of the implementation, not necessarily part of the spec, since there will be both Go and Javascript implementations of this functionality, each will have its own appropriate API.

Reviewer 2: If signer S signs msg M, is that signature always the same, or does it depend on the "renderer" of that msg?

Reviewer 4: The signature S of a Msg M depends on the signing mode (Textual, Direct, Legacy Amino, Legacy Amino JSON) chosen for the transaction, and even more basically it depends on the particular blockchain in question, since some Msg types will be supported on some blockchains and not on others. So for a particular revision of software for a particular blockchain, and a particular signing mode for a particular message, then yes, that must determine a sequence of bytes to feed into the signing algorithm for signature and for verification.

The feature discussed in this PR allows a particular revision of a particular blockchain to customize the default SDK behavior when using the Textual mode, but once it has done so, it must be deterministic for a given Msg.

Reviewer 2: Ah, so a given transaction on a given blockchain at a given height will always have precisely one signing mode? Great!

Reviewer 4: No - given all the preceding *and* the signing mode, it will have precisely representation that is input to signing. For a given blockchain and txn you can choose to sign either the text mode translation, or the JSON mode translation, or whatever.

Reviewer 2: So that's a big problem :) 

A tx/msg is Merkelized as a sequence of bytes stored on chain. If you sign anything other than that specific sequence of bytes, then you're not signing the tx/msg itself, you're signing a transformation. The transform function isn't part of consensus, so you can't assume sign_mode_textual will always do the same thing. And it's basically not possible to assert that a transformation of one byte sequence to another is equivalent and lossless at a semantic (i.e. the protobuf) level, so you can't assume that verified signatures in different modes have actually signed the same information and establish equivalent trust.

IMO this is a critical, SEV-0, red-alert kind of problem. But take that feedback as you like 😉

Reviewer 4: Your feedback seems like it applies to Sign Mode Textual as a whole, not the specific changes contemplated in this PR, so I'll ask you to take your concerns to a more appropriate venue. Perhaps Discord.

```
// for renderer, msg, and ctx of the right type
keyvals, err := renderer.Format(ctx, msg)
if err != nil {
    fail_check()
}
msg2, err := renderer.Parse(ctx, keyvals)
if err != nil {
    fail_check()
}
if !proto.Equal(msg, msg2) {
    fail_check()
}
pass_check()

> ```

Reviewer 4: Perhaps custom renderers and the default renderers should just use the same interface? We'd have to smooth out the differences. E.g. separate types for renderers that just fill out the "vallue" part of the key/value vs those that emit multiple key/value lines?
